pkg/globalnet/controllers/ipam/ippool_test.go

```diff
@@ -0,0 +1,261 @@
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		It("should return the requested IP", func() {
+			Expect(service1Ip).Should(Equal(requestIp1))
+		})
+
+		When("the same IP but different key is requested", func() {
+			requestedIp2, err := pool.RequestIp(pod1, requestIp1)
+
+			It("should return a different IP", func() {
+				Expect(err).NotTo(HaveOccurred())
+				Expect(requestedIp2).ShouldNot(Equal(requestIp1))
+			})
+		})
+
+		When("the same IP and key is requested", func() {
+			requestedIp2, err := pool.RequestIp(service1, requestIp1)
+
+			It("should return same IP", func() {
+				Expect(err).NotTo(HaveOccurred())
+				Expect(requestedIp2).Should(Equal(requestIp1))
+			})
+		})
+	})
+
+	When("All IPs are allocated", func() {
+		pool, _ := NewIpPool(testCidr)
+		service1Ip, err := pool.Allocate(service1)
+		_, err = pool.Allocate(pod1)
+
+		It("should not return an error", func() {
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		When("IsFull is called", func() {
+			It("should return true", func() {
+				Expect(pool.IsFull()).To(BeTrue())
+			})
+		})
+
+		When("any IP is requested for an unallocated key", func() {
+			pod2Ip, err := pool.Allocate(pod2)
+
+			It("should return an error", func() {
+				Expect(err).To(HaveOccurred())
+				Expect(pod2Ip).Should(Equal(""))
+			})
+		})
+
+		When("the same IP is requested for an allocated key", func() {
+			requestedIp, err := pool.RequestIp(service1, service1Ip)
+
+			It("should not return an error", func() {
+				Expect(err).NotTo(HaveOccurred())
+			})
+
+			It("should return same IP", func() {
+				Expect(requestedIp).Should(Equal(service1Ip))
+			})
+		})
+
+		When("a different IP is requested for an allocated key", func() {
+			requestedIp, err := pool.RequestIp(pod1, requestIp1)
+
+			It("should return an error", func() {
+				Expect(err).To(HaveOccurred())
+				Expect(requestedIp).Should(Equal(""))
+			})
+		})
+	})
+}
+
+func testIpRelease() {
+	When("Allocated IP is Released", func() {
+		pool, _ := NewIpPool(testCidr)
+		service1Ip, _ := pool.Allocate(service1)
+		available := len(pool.available)
+		allocated := len(pool.allocated)
+		releasedIp := pool.Release(service1)
+
+		It("should return IP released", func() {
+			Expect(releasedIp).Should(Equal(service1Ip))
+		})
+
+		It("should increment available IP count by 1", func() {
+			Expect(len(pool.available) - available).Should(Equal(1))
+		})
+
+		It("should decrement allocated IP count by 1", func() {
+			Expect(allocated - len(pool.allocated)).Should(Equal(1))
+		})
+
+		It("should become available for allocation", func() {
+			Expect(pool.IsAvailable(releasedIp)).To(BeTrue())
+		})
+	})
+
+	When("Unallocated IP is Released", func() {
+		pool, _ := NewIpPool(testCidr)
+		_, err := pool.Allocate(service1)
+		available := len(pool.available)
+		allocated := len(pool.allocated)
+		releasedIp := pool.Release(pod1)
+
+		It("should not return an error", func() {
+			Expect(err).NotTo(HaveOccurred())
+		})
```

Comment:
The err we are validating here corresponds to Allocate API. Release() does not return err, so IMO this error validation can be removed. 

Other than this, everything looks good. Nice work!! 
Please squash the commits and we can merge the PR.

Reply:
I get lint errors if I don't store error and then not use it somewhere. That is why had to add it.

Reply:
Okay. Please squash the commits.